### PR TITLE
Cc 0009 sign in by using email and password

### PR DESCRIPTION
**Pull request by VanderDeng**
Sign in page
image

image

### CODE REVIEW

File: src/__tests__/components/SignIn/SignIn.test.tsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    user.click(submitBtn);
+    await waitFor(() => expect(screen.getByRole("status")).toBeInTheDocument());
+  });
+
+  const mock = new MockAdapter(customAxios, { onNoMatch: "throwException" });
+  const account = { username: "vvv@v.com", password: "123qweASD=" };
+
+  beforeAll(() => {
+    mock.reset();
+  });
```
[the remark this answers is not in the document]
Comment (kritomomo) [reply src/__tests__/components/SignIn/SignIn.test.tsx:43]: Or you can write another describe for this integration test

File: src/__tests__/components/SignIn/SignIn.test.tsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+  it("should call success", async () => {
+    mock.onPost("/staff/signin", account).reply(200, account);
+    const { result } = renderHook(() => userSignIn());
+
+    act(() => {
+      result.current.handleSignInSubmit(account);
+    });
+    await waitFor(() => expect(result.current.isLoading).toBe(true));
```
Comment (kritomomo) [inline src/__tests__/components/SignIn/SignIn.test.tsx:52]: You didn't expect loading false because of that warning?

File: src/__tests__/components/SignIn/SignIn.test.tsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  });
+
+  it("should call error", async () => {
+    mock.onPost("/staff/signin", account).reply(400, account);
```
Comment (kritomomo) [inline src/__tests__/components/SignIn/SignIn.test.tsx:56]: image
There is no 400 status.

File: src/__tests__/components/SignIn/SignIn.test.tsx [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    const { result } = renderHook(() => userSignIn());
+
+    act(() => {
+      result.current.handleSignInSubmit(account);
+    });
+    await waitFor(() => expect(result.current.isLoading).toBe(true));
```
Comment (kritomomo) [inline src/__tests__/components/SignIn/SignIn.test.tsx:62]: Do you need to add a `expect` for toast?

File: src/components/SignIn/useSignIn.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    return response;
+  };
+  const handleSignInSubmit = async (data: { username: string; password: string }) => {
+    setIsLoading(true);
+    try {
+      const response = await signInRequest(data.username, data.password);
+      if (response.status === 200) {
+        router.push("/");
+      }
+    } catch (error: any) {
```
Comment (kritomomo) [inline src/components/SignIn/useSignIn.ts:24]: If you can, try to avoid using any

File: src/components/SignIn/useSignIn.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+      toast({
+        title: "sign in failed",
```
Comment (kritomomo) [inline src/components/SignIn/useSignIn.ts:26]: The first letter of the title should be capitalized

File: src/components/SignIn/useSignIn.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        description: "username and password is not authenticated",
```
Comment (kritomomo) [inline src/components/SignIn/useSignIn.ts:27]: image
Same

**Comment by kritomomo** [comment]
You need to add a UI image to the description

File: src/pages/_document.tsx [hunk 8]
```diff
@@ -1,61 +1,33 @@
...
+    // Retrieve styles from components in the page
+    const originalRenderPage = ctx.renderPage;
+
+    try {
+      ctx.renderPage = () =>
+        originalRenderPage({
+          enhanceApp: (App) => (props) => sheet.collectStyles(<App {...props} />),
+        });
+
+      const initialProps = await Document.getInitialProps(ctx);
```
[the remark this answers is not in the document]
Comment (VanderDeng) [reply src/pages/_document.tsx:18]: solved this missed conflict

File: src/__tests__/components/SignIn/SignIn.test.tsx [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    expect(screen.getByText("CourtCanva")).toBeInTheDocument();
+    expect(screen.getByText("Sign in to CourtCanva Franchisee")).toBeInTheDocument();
+  });
+
+  it("should show warning", async () => {
+    renderWithMockedProvider(<SignIn />);
+    const submitBtn = screen.getByText("Sign In");
+    user.click(submitBtn);
+
+    await waitFor(() => expect(screen.getByText("Username is required")).toBeInTheDocument());
```
Comment (yuanningd) [inline src/__tests__/components/SignIn/SignIn.test.tsx:23]: you can use `findByText` here

File: src/components/SignIn/useSignIn.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    const response = await axios.post("/staff/signin", {
+      username,
+      password,
+    });
+    return response;
+  };
+  const handleSignInSubmit = async (data: { username: string; password: string }) => {
+    setIsLoading(true);
+    try {
+      const response = await signInRequest(data.username, data.password);
```
[the remark this answers is not in the document]
Comment (VanderDeng) [reply src/components/SignIn/useSignIn.ts:20]: ok, I delete the function signInRequest

File: src/components/SignIn/useSignIn.ts [hunk 11]
```diff
@@ -0,0 +1,38 @@
+import axios from "../../utils/axios";
+import { useState } from "react";
+import { useRouter } from "next/router";
+import { useToast } from "@chakra-ui/react";
+
+export default function useSignIn() {
+  const [isLoading, setIsLoading] = useState<boolean>(false);
```
Comment (yuanningd) [inline src/components/SignIn/useSignIn.ts:7]: `<boolean>` is unnecessary since it can be inferred from your initial value `false`.

File: src/components/SignIn/useSignIn.ts (hunk 10 above)
Comment (yuanningd) [inline src/components/SignIn/useSignIn.ts:20]: You have deconstructed `username` and `password` from `data`, why not use them directly

File: src/components/SignIn/index.tsx [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+          </Flex>
+          <Flex>
+            <SignInForm />
+          </Flex>
+        </Flex>
+        <Flex width="40%">
+          <Image
+            height="100vh"
+            objectFit="cover"
+            src="./signIn/signIn-image.png"
```
[the remark this answers is not in the document]
Comment (VanderDeng) [reply src/components/SignIn/index.tsx:31]: the problem is chakra <Image> cannot recognize the image path if I put the image in the src/assets.
https://stackoverflow.com/questions/68339712/chakra-ui-image-component-not-recognizing-image-path

File: src/components/SignIn/useSignIn.ts [hunk 13]
```diff
@@ -0,0 +1,38 @@
+import axios from "../../utils/axios";
+import { useState } from "react";
+import { useRouter } from "next/router";
+import { useToast } from "@chakra-ui/react";
+
+export default function useSignIn() {
```
Comment (yuanningd) [inline src/components/SignIn/useSignIn.ts:6]: Is there any specific reason for you to extract a custom hook here? Since it is not reusable, you must have a good reason to do so.

File: src/components/SignIn/index.tsx [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+            alt="signIn-image"
```
[the remark this answers is not in the document]
Comment (VanderDeng) [reply src/components/SignIn/index.tsx:32]: I've changed this one

File: src/__tests__/components/SignIn/SignIn.test.tsx (hunk 3 above)
Comment (VanderDeng) [reply src/__tests__/components/SignIn/SignIn.test.tsx:56]: done

File: src/components/SignIn/useSignIn.ts (hunk 6 above)
Comment (VanderDeng) [reply src/components/SignIn/useSignIn.ts:26]: done

File: src/components/SignIn/useSignIn.ts (hunk 7 above)
Comment (VanderDeng) [reply src/components/SignIn/useSignIn.ts:27]: done

File: src/__tests__/components/SignIn/SignIn.test.tsx (hunk 9 above)
Comment (VanderDeng) [reply src/__tests__/components/SignIn/SignIn.test.tsx:23]: done

File: src/components/SignIn/useSignIn.ts (hunk 11 above)
Comment (VanderDeng) [reply src/components/SignIn/useSignIn.ts:7]: done

File: src/components/SignIn/useSignIn.ts (hunk 10 above)
Comment (VanderDeng) [reply src/components/SignIn/useSignIn.ts:20]: done

File: src/components/SignIn/useSignIn.ts (hunk 5 above)
Comment (VanderDeng) [reply src/components/SignIn/useSignIn.ts:24]: done

File: src/components/SignIn/SignInForm.tsx [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+
+  return (
+    <form style={{ height: "400px", width: "360px" }} onSubmit={handleSubmit(formSubmit)}>
+      <FormLabel fontWeight="600">Username</FormLabel>
+      <Input
+        placeholder="Enter username"
+        role="username"
+        type="email"
+        {...register("username", {
+          required: "Username is required",
```
[the remark this answers is not in the document]
Comment (VanderDeng) [reply src/components/SignIn/SignInForm.tsx:31]: name is from Figma design

File: src/components/SignIn/SignInForm.tsx [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+};
+
+const SignInForm = () => {
+  const { isLoading, handleSignInSubmit } = useSignIn();
+  const {
+    register,
+    formState: { errors },
+    handleSubmit,
+  } = useForm<FormData>();
+  const formSubmit = (data: { username: string; password: string }) => {
```
[the remark this answers is not in the document]
Comment (VanderDeng) [reply src/components/SignIn/SignInForm.tsx:19]: done

File: src/__tests__/components/SignIn/SignIn.test.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (VanderDeng) [reply src/__tests__/components/SignIn/SignIn.test.tsx:43]: done

File: src/__tests__/components/SignIn/SignIn.test.tsx [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+import { act } from "react-dom/test-utils";
+
+describe("Sign in Page", () => {
+  it("should render personal sign in page success", () => {
+    renderWithMockedProvider(<SignIn />);
+    expect(screen.getByText("CourtCanva")).toBeInTheDocument();
+    expect(screen.getByText("Sign in to CourtCanva Franchisee")).toBeInTheDocument();
+  });
+
+  it("should show warning", async () => {
```
[the remark this answers is not in the document]
Comment (VanderDeng) [reply src/__tests__/components/SignIn/SignIn.test.tsx:18]: done

File: src/__tests__/components/SignIn/SignIn.test.tsx [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+    await waitFor(() => expect(screen.getByText("Username is required")).toBeInTheDocument());
+    await waitFor(() => expect(screen.getByText("Password is required")).toBeInTheDocument());
+  });
+
+  it("should form submit", async () => {
+    renderWithMockedProvider(<SignIn />);
+    const submitBtn = screen.getByText("Sign In");
+    const usernameInput = screen.getByRole("username");
+    const passwordInput = screen.getByRole("password");
+    fireEvent.change(usernameInput, { target: { value: "123@123.com" } });
```
[the remark this answers is not in the document]
Comment (VanderDeng) [reply src/__tests__/components/SignIn/SignIn.test.tsx:32]: done, I change it to userevent

File: src/__tests__/components/SignIn/SignIn.test.tsx (hunk 2 above)
Comment (VanderDeng) [reply src/__tests__/components/SignIn/SignIn.test.tsx:52]: done, I fixed it, though there is still a warning
image

File: src/__tests__/components/SignIn/SignIn.test.tsx (hunk 4 above)
Comment (VanderDeng) [reply src/__tests__/components/SignIn/SignIn.test.tsx:62]: done, I add a test for toast after return error

File: src/components/SignIn/useSignIn.ts (hunk 13 above)
Comment (VanderDeng) [reply src/components/SignIn/useSignIn.ts:6]: to manage all the API

File: src/components/SignIn/SignInForm.tsx [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+    handleSignInSubmit(data);
+  };
+
+  return (
+    <form style={{ height: "400px", width: "360px" }} onSubmit={handleSubmit(formSubmit)}>
```
[the remark this answers is not in the document]
Comment (VanderDeng) [reply src/components/SignIn/SignInForm.tsx:24]: done

File: src/components/SignIn/index.tsx [hunk 20]
```diff
@@ -0,0 +1,40 @@
+import React from "react";
+import HeaderLayout from "../../layouts/HeaderLayout";
+import SignInForm from "./SignInForm";
+
+import { Flex, Image, Text } from "@chakra-ui/react";
+
+const SignIn: React.FC = () => {
+  return (
+    <HeaderLayout>
```
[the remark this answers is not in the document]
Comment (VanderDeng) [reply src/components/SignIn/index.tsx:9]: done

File: src/components/SignIn/SignInForm.tsx [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+        role="username"
+        type="email"
+        {...register("username", {
+          required: "Username is required",
+        })}
+      />
+      <ErrorMessage
+        errors={errors}
+        name="username"
+        render={({ message }) => <p style={{ color: "red" }}>{message}</p>}
```
[the remark this answers is not in the document]
Comment (VanderDeng) [reply src/components/SignIn/SignInForm.tsx:37]: done

File: src/components/SignIn/index.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (kritomomo) [reply src/components/SignIn/index.tsx:31]: public/assets

[the remark this answers is not in the document]
Comment (VanderDeng) [reply src/components/SignIn/index.tsx]: done

File: src/components/SignIn/useSignIn.ts (hunk 13 above)
Comment (kritomomo) [reply src/components/SignIn/useSignIn.ts:7]: It is true that building customHooks can let you extract component logic into reusable functions. you can also separate logic and UI to improve the reading experience (of course, this varies from person to person, some people think that if this logic is only used in this component, then Just put it in there),
In addition, another reason to separate the logic is to use renderHook (from @testing-library/react) to test this logic separately during testing

File: src/components/SignIn/index.tsx [hunk 22]
```diff
@@ -0,0 +1,44 @@
+import React from "react";
+import HeaderLayout from "../../layouts/HeaderLayout";
+import SignInForm from "./SignInForm";
+
+import { Flex, Image, Text } from "@chakra-ui/react";
+
+const SignIn = () => {
+  return (
+    <HeaderLayout title="Sign In">
```
Comment (tatoMa) [inline src/components/SignIn/index.tsx:9]: Head tag should be on the page file cos it is not related to any component.

File: src/components/SignIn/SignInForm.tsx [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+      <Input
+        placeholder="Enter password"
+        role="password"
+        type="password"
+        {...register("password", {
+          required: "Password is required",
+        })}
+      />
+      <Container color="red">
+        <ErrorMessage errors={errors} name="password" render={({ message }) => <p>{message}</p>} />
```
Comment (tatoMa) [inline src/components/SignIn/SignInForm.tsx:49]: The function inside of these 2 renders in ErrorMessage can be extracted to a standalone one. Makes this code DRYer.

File: src/components/SignIn/index.tsx [hunk 24]
```diff
@@ -0,0 +1,44 @@
+import React from "react";
```
Comment (tatoMa) [inline src/components/SignIn/index.tsx:1]: No need to import the React from all files in a Nextjs project.

**Review by tatoMa — COMMENTED** [review]
Nice work overall.

File: src/components/SignIn/index.tsx (hunk 22 above)
Comment (VanderDeng) [reply src/components/SignIn/index.tsx:9]: done

File: src/components/SignIn/index.tsx (hunk 24 above)
Comment (VanderDeng) [reply src/components/SignIn/index.tsx:1]: done

File: src/components/SignIn/SignInForm.tsx (hunk 23 above)
Comment (VanderDeng) [reply src/components/SignIn/SignInForm.tsx]: done

**Review by tatoMa — APPROVED** [review]

File: src/__tests__/components/SignIn/SignIn.test.tsx [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+
+  it("should form submit", async () => {
+    renderWithMockedProvider(<SignIn />);
+    const submitBtn = screen.getByText("Sign In");
+    const usernameInput = screen.getByRole("email");
+    const passwordInput = screen.getByRole("password");
+    await user.type(usernameInput, account.username);
+    await user.type(passwordInput, account.password);
+    user.click(submitBtn);
+    await waitFor(() => expect(screen.getByRole("status")).toBeInTheDocument());
```
Comment (yuanningd) [inline src/__tests__/components/SignIn/SignIn.test.tsx:40]: all the combinations of `waitFor` and `getBy...` can be replaced by `findBy`

**Review by yuanningd — APPROVED** [review]

File: src/__tests__/components/SignIn/SignIn.test.tsx (hunk 25 above)
Comment (VanderDeng) [reply src/__tests__/components/SignIn/SignIn.test.tsx]: done
